Add FreeBSD support for ZFS datasets and pools

jbenden opened the pull request:
The following new resources have been added; however, they presently only support FreeBSD and similar.

* `zfs_dataset`: tests if a named ZFS dataset is present and/or has certain properties.
* `zfs_pool`: tests if a named ZFS pool is present and/or has certain properties.

The `mount` resource has been refactored to include support for FreeBSD too.

adamleff commented on lib/resources/mount.rb:
```diff
@@ -49,12 +49,38 @@ def method_missing(name)
+  end
+
+  class LinuxMounts < MountsInfo
+    include MountParser
```
This is probably a good time to rename this to `LinuxMountParser` to avoid any confusion.

adamleff commented on lib/resources/mount.rb:
```diff
@@ -49,12 +49,38 @@ def method_missing(name)
+    def initialize(inspec)
+      @inspec = inspec
+    end
+  end
```
Should we add a default `#parse_mount_options` method here that `raise`s an exception to make it clear for future users that the method is required?

adamleff commented on lib/resources/zfs_dataset.rb:
```diff
@@ -0,0 +1,62 @@
+    "
+
+    def initialize(zfs_dataset)
+      return skip_resource 'The `zfs_dataset` resource is not supported on your OS yet.' if !inspec.os.bsd?
```
We don't need to account for it in this PR, but I suspect this resource would work OK as-is on Solaris and any Linux that is using ZFS-on-Linux too, with the small exception of the actual ZFS binary paths may be different. This is worth some investigation.

chris-rock replied:
Would love to see the Solaris support for this as well!

adamleff reviewed (changes requested):
@jbenden Thanks for your contribution! I'm a huge fan of ZFS... used it a lot on OmniOS at my last job... so adding ZFS support fills me with glee. :)

I added a few comments, and we should add some unit tests to this with some mocked output for all the items changed/added (i.e. BSD mounts, ZFS pools, and ZFS datasets). I'd be happy to help you navigate that process if you need any assistance.

jbenden committed:
Add FreeBSD support for ZFS datasets and pools

The following new resources have been added; however, they
presently only support FreeBSD and similar.

* `zfs_dataset`: tests if a named ZFS dataset is present
  and/or has certain properties.
* `zfs_pool`: tests if a named ZFS pool is present and/or
  has certain properties.

Additionally, the `mount` resource has been reworked to
include support for FreeBSD; while the existing class
was renamed to LinuxMountParser.

Unit-tests were added for all of the above.

Signed-off-by: Joseph Benden <joe@benden.us>
